sisyphus/prowAccessor.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sisyphus
+
+import (
+	"encoding/json"
+	"log"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	u "istio.io/test-infra/toolbox/util"
+)
+
+const (
+	lastBuildTXT = "latest-build.txt"
+	finishedJSON = "finished.json"
+	startedJSON  = "started.json"
+)
+
+// IProwAccessor defines the Prow interface
+type IProwAccessor interface {
+	GetLatestRun(jobName string) (int, error)
+	GetProwResult(jobName string, runNo int) (*ProwResult, error)
+	Rerun(jobName string, runNo, numRerun int) error
+	GetGubernatorURL() string
+}
+
+// ProwResult matches the structure published in finished.json
+type ProwResult struct {
+	TimeStamp  uint32       `json:"timestamp"`
+	Version    string       `json:"version"`
+	Result     string       `json:"result"`
+	Passed     bool         `json:"passed"`
+	JobVersion string       `json:"job-version"`
+	Metadata   ProwMetadata `json:"metadata"`
+}
+
+// ProwMetadata matches the structure published in finished.json
+type ProwMetadata struct {
+	Repo       string `json:"repo"`
+	RepoCommit string `json:"repo-commit"`
+}
+
+// ProwJobConfig matches the structure published in started.json
+type ProwJobConfig struct {
+	Node        string `json:"node"`
+	JenkinsNode string `json:"jenkins-node"`
+	Version     string `json:"version"`
+	TimeStamp   uint32 `json:"timestamp"`
+	RepoVersion string `json:"repo-version"`
+}
+
+// ProwAccessor provides programmable access to Prow data on GCS
+type ProwAccessor struct {
+	prowProject   string
+	prowZone      string
+	gubernatorURL string
+	gcsClient     *u.GCSClient
+}
+
+// NewProwAccessor creates a new ProwAccessor
+func NewProwAccessor(prowProject, prowZone, gubernatorURL, gcsBucket string) *ProwAccessor {
+	return &ProwAccessor{
+		prowProject:   prowProject,
+		prowZone:      prowZone,
+		gubernatorURL: gubernatorURL,
+		gcsClient:     u.NewGCSClient(gcsBucket),
```

Comment:
Instead of using GCSClient, you should use an interface that just Read(). That way instead of mocking ProwAccessor you just need to mock the storage, and you can actually test the ProwAccessor code.

Reply:
Done